lib/dice/src/cert.rs

```diff
@@ -0,0 +1,340 @@
+    const ISSUER_SN_END: usize = alias_cert_tmpl::ISSUER_SN_END;
+    const NOTBEFORE_START: usize = alias_cert_tmpl::NOTBEFORE_START;
+    const NOTBEFORE_END: usize = alias_cert_tmpl::NOTBEFORE_END;
+    const SUBJECT_SN_START: usize = alias_cert_tmpl::SUBJECT_SN_START;
+    const SUBJECT_SN_END: usize = alias_cert_tmpl::SUBJECT_SN_END;
+    const PUB_START: usize = alias_cert_tmpl::PUB_START;
+    const PUB_END: usize = alias_cert_tmpl::PUB_END;
+    const SIG_START: usize = alias_cert_tmpl::SIG_START;
+    const SIG_END: usize = alias_cert_tmpl::SIG_END;
+    const SIGNDATA_START: usize = alias_cert_tmpl::SIGNDATA_START;
+    const SIGNDATA_END: usize = alias_cert_tmpl::SIGNDATA_END;
+
+    fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    fn as_mut_bytes(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+impl AliasCert {
+    pub fn set_fwid(
+        mut self,
+        fwid: &[u8; alias_cert_tmpl::FWID_LENGTH],
+    ) -> Self {
+        self.0[alias_cert_tmpl::FWID_START..alias_cert_tmpl::FWID_END]
+            .copy_from_slice(fwid);
+
+        self
+    }
+
+    pub fn get_fwid(&self) -> &[u8; alias_cert_tmpl::FWID_LENGTH] {
+        self.0[alias_cert_tmpl::FWID_START..alias_cert_tmpl::FWID_END]
+            .try_into()
+            .unwrap_lite()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use core::str::FromStr;
+
+    #[test]
+    fn serial_number_from_new() {
+        let sn: u8 = 0x10;
+        let cert = DeviceIdSelfCert([0u8; deviceid_cert_tmpl::SIZE])
+            .set_serial_number(sn);
+
+        assert_eq!(sn, cert.get_serial_number());
+    }
+
+    #[test]
+    fn issuer_sn_from_new() {
+        let sn = SerialNumber::from_str("0123456789ab").expect("SN from_str");
+        let cert = DeviceIdSelfCert([0u8; deviceid_cert_tmpl::SIZE])
+            .set_issuer_sn(&sn);
+
+        assert_eq!(cert.get_issuer_sn().as_bytes(), sn.as_bytes());
+    }
+
+    #[test]
+    fn subject_sn_from_new() {
+        let sn = SerialNumber::from_str("0123456789ab").expect("SN from_str");
+        let cert = DeviceIdSelfCert([0u8; deviceid_cert_tmpl::SIZE])
+            .set_subject_sn(&sn);
+
+        assert_eq!(cert.get_subject_sn().as_bytes(), sn.as_bytes());
+    }
+
+    // Signature over CERT with issuer / subject SN & PUBKEY set according
+    // to 'sign' test below.
+    const SIG_EXPECTED: [u8; SIGNATURE_SERIALIZED_LENGTH] = [
+        0x26, 0x2A, 0x81, 0xE9, 0x1F, 0x06, 0xCF, 0xF0, 0x13, 0xEB, 0x33, 0x71,
+        0x5A, 0xB9, 0x5C, 0xC0, 0xC7, 0x40, 0x01, 0x83, 0x7C, 0xB6, 0x2F, 0x2E,
+        0x88, 0xE9, 0x95, 0xD9, 0x10, 0x9C, 0xD8, 0xF5, 0x33, 0x4B, 0x9B, 0xB1,
+        0x6A, 0xB3, 0x23, 0xDB, 0x3A, 0x1C, 0x35, 0x31, 0xE7, 0x38, 0xEC, 0x9B,
+        0xAA, 0x32, 0x36, 0x5A, 0xAA, 0x37, 0x4B, 0xF5, 0xE7, 0x7A, 0x2C, 0x4E,
+        0x88, 0x35, 0x50, 0x0E,
+    ];
+
+    // Each time this library is built the notBefore time in the validity
```

Comment:
I had initially set this up such that the cert templates were generated at build time. I pulled this out because it had weird dependencies on external tools used to create the x.509 structures. In constructing the templates I was initially setting the notBefore field with the system time before I realized that this would break reproducible builds. This can be done in such a way it won't break reproducible builds but it requires support from xtask. I'll make a ticket and add a TODO.

Reply:
https://github.com/oxidecomputer/hubris/issues/752